custom_components/divoom_pixoo/pages/progress_bar.py

```diff
@@ -7,57 +7,71 @@
 
 _LOGGER = logging.getLogger(__name__)
 
-def progress_bar(pixoo, hass, page_data: dict, FONT_PICO_8, FONT_GICKO, FIVE_PIX, ELEVEN_PIX, CLOCK):
+
+def progress_bar(
+    pixoo, hass, page_data: dict, FONT_PICO_8, FONT_GICKO, FIVE_PIX, ELEVEN_PIX, CLOCK
+):
     pixoo.clear()
 
-    #Datetime
+    # Datetime
     now = datetime.now()
     time = now.strftime("%H:%M")
+    # Custom time_end, if provided
+    if time_end := page_data.get("time_end", ""):
+        time_end = (
+            datetime.fromisoformat(time_end)
+            .astimezone(now.astimezone().tzinfo)
+            .strftime("%H:%M")
+        )
 
```

Comment:
time_end isn't rendered!!!!! Templates won't work in the config (List of Pages)!!!

Reply:
im not sure what you mean? I've tested this and it is working in the configuration.

Reply:
```
2024-03-30 16:17:36 2024-03-30 20:17:36.764 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up divoom_pixoo platform for sensor
2024-03-30 16:17:36 Traceback (most recent call last):
2024-03-30 16:17:36   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 352, in _async_setup_platform
2024-03-30 16:17:36     await asyncio.gather(*pending)
2024-03-30 16:17:36   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 533, in async_add_entities
2024-03-30 16:17:36     await asyncio.gather(*tasks)
2024-03-30 16:17:36   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 782, in _async_add_entity
2024-03-30 16:17:36     await entity.add_to_platform_finish()
2024-03-30 16:17:36   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1324, in add_to_platform_finish
2024-03-30 16:17:36     await self.async_added_to_hass()
2024-03-30 16:17:36   File "/config/custom_components/divoom_pixoo/sensor.py", line 87, in async_added_to_hass
2024-03-30 16:17:36     await self._async_update()
2024-03-30 16:17:36   File "/config/custom_components/divoom_pixoo/sensor.py", line 97, in _async_update
2024-03-30 16:17:36     await self._async_next_page()
2024-03-30 16:17:36   File "/config/custom_components/divoom_pixoo/sensor.py", line 125, in _async_next_page
2024-03-30 16:17:36     await self.hass.async_add_executor_job(self._render_page, self.page)
2024-03-30 16:17:36   File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
2024-03-30 16:17:36     result = self.fn(*self.args, **self.kwargs)
2024-03-30 16:17:36              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-03-30 16:17:36   File "/config/custom_components/divoom_pixoo/sensor.py", line 136, in _render_page
2024-03-30 16:17:36     special_pages[page_type](pixoo, self.hass, page)
2024-03-30 16:17:36   File "/config/custom_components/divoom_pixoo/pages/_pages.py", line 10, in <lambda>
2024-03-30 16:17:36     "progress_bar": lambda pixoo, hass, page: progress_bar(pixoo, hass, page, FONT_PICO_8, FONT_GICKO, FIVE_PIX, ELEVEN_PIX, CLOCK),
2024-03-30 16:17:36                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-03-30 16:17:36   File "/config/custom_components/divoom_pixoo/pages/progress_bar.py", line 22, in progress_bar
2024-03-30 16:17:36     datetime.fromisoformat(time_end)
2024-03-30 16:17:36 ValueError: Invalid isoformat string: "{{states('input_text.aaaa')}}"
```

Here's the config:
![image](https://github.com/gickowtf/pixoo-homeassistant/assets/45701824/c0455324-c83a-43b9-b4dc-a7ec030863f8)

By using the same page/data in the service, it will because, because it's Home Assistant that renders the template. But in the config, it has to be done ourselves.

Reply:
Ah ok, I should wrap this to detect if it's a datetime string first?

Which is what it is expecting? ( I don't know if it's possible to detect the underlying type from page_data)

I know I ran into errors before with passing non-numbers to `percentage:`, but maybe that's more obvious required as an `int` or is part of the documentation - which I forgot to include in this PR

Reply:
OHH Are you referring to the Template(....) call like https://github.com/yyolk/pixoo-homeassistant/blob/b5095050ddf531f3515bc0b47ef1d83c9b5079a3/custom_components/divoom_pixoo/pages/progress_bar.py#L33-L36

Should I wrap it like these?

Reply:
> Should I wrap it like these?

Exactly!

> Sorry for misunderstanding, thank you for your patience with this being my first custom integration I've worked on! 🙇

No worries. This integration also isn't like others 😉

Reply:
Thank you for your prompt feedback! 

I think my disconnect was that I wasn't rendering any template in the configuration (i used `{{ states.sensor.work_day.state }}`), so I didn't know it would be an issue. I also didn't run the 8h timer very long, but it showed what appeared to be the correct delta- so I learned today this and not everything in a `{{ ... }}` is considered a template and is evaluated differently

* * * 

With the latest change: 

So I don't know if I should make `time_end` less complex than what I'm doing in this latest change. 

I went with catching if it's an invalid isoformat string while evaluating the template, which is also within the try..except of the other template evaluations - and it should only run if something is set there.

I tried this template:

```yaml
time_end: "{{ now() + timedelta(hours=3) }}"
```